src/dsa/rpo_falcon512/signature.rs

```diff
@@ -1,35 +1,38 @@
-use alloc::string::ToString;
-use core::cell::OnceCell;
+use alloc::{string::ToString, vec::Vec};
+use core::ops::Deref;
 
 use super::{
-    ByteReader, ByteWriter, Deserializable, DeserializationError, Felt, NonceBytes, NonceElements,
-    Polynomial, PublicKeyBytes, Rpo256, Serializable, SignatureBytes, Word, MODULUS, N,
-    SIG_L2_BOUND, ZERO,
+    error::FalconSerializationError,
+    hash_to_point::hash_to_point_rpo256,
+    keys::PubKeyPoly,
+    math::{FalconFelt, FastFft, Polynomial},
+    ByteReader, ByteWriter, Deserializable, DeserializationError, Felt, Nonce, Rpo256,
+    Serializable, Word, LOG_N, MODULUS, N, SIG_L2_BOUND, SIG_POLY_BYTE_LEN,
 };
+use num::Zero;
 
 // FALCON SIGNATURE
 // ================================================================================================
 
 /// An RPO Falcon512 signature over a message.
 ///
-/// The signature is a pair of polynomials (s1, s2) in (Z_p\[x\]/(phi))^2, where:
+/// The signature is a pair of polynomials (s1, s2) in (Z_p\[x\]/(phi))^2 and a nonce `r`,
+/// where:
 /// - p := 12289
 /// - phi := x^512 + 1
-/// - s1 = c - s2 * h
-/// - h is a polynomial representing the public key and c is a polynomial that is the hash-to-point
-///   of the message being signed.
 ///
 /// The signature  verifies if and only if:
-/// 1. s1 = c - s2 * h
+/// 1. h = s2^(-1) * (c - s1)
 /// 2. |s1|^2 + |s2|^2 <= SIG_L2_BOUND
 ///
-/// where |.| is the norm.
+/// where |.| is the norm and:
+/// - pk = Rpo256::hash(h)
+/// - c = HashToPoint(r || message)
```

Comment:
Is this still representative of what we do below? (i.e., first recover public and then verify the signature in the canonical way).

Reply:
Updated